### PR TITLE
First stage of tiered JIT infrastructure: Interpreter tier

### PR DESCRIPTION
**Pull request by CrystalGamma**
This PR introduces a new framework for JITs (though no code generation is implemented in this PR, a PoC will be coming soon hopefully). It separates the JitBase interface from various implementation details that are incompatible with concurrent compilation, which is planned as the major improvements offered by the new framework.

This PR includes a generic code path which is similar in functionality to CachedInterpreter (and reaches around 50% to 110% of its performance, depending on workload and system) and offers the kinds of interfaces that can be used by a concurrent compiler (it supplies code blocks with various annotations, like if non-RAM addresses were accessed, and I$ invalidations in a 'report' data structure, since the compiler cannot access guest memory, and has provisions for non-interpreter blocks in its dispatch cache).

Plans for the following PRs are to create a PoC Baseline JIT for x86, which will not do register caching yet and only implement the ~5 most frequent instructions (`addi`, `bc`, `lwz`, `stw`, potentially `cmpl` and `rlwinm`) and call Interpreter for all others [first following PR, currently working on it], create a PoC Baseline JIT for POWER with a similar feature set [second PR], implement 10–20 additional instructions and register caching for both architectures [third PR] and after that, start looking at JIT tiers with cross-block optimizations.

The first commit in this PR (which is responsible for a lot of the changed files) is actually from #7602 which has been approved, but not merged at the time of writing. I am unsure of if the commit history is OK like this … I am very amenable to requests to squash some of the last 7 commits into others, but don't know if it would be useful.

### CODE REVIEW

File: Source/Core/Common/HandShake.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include <mutex>
+#include <optional>
+
+template <typename Inner>
+struct HandShake
+{
+public:
+  class ReaderGuard
+  {
+  private:
```
Comment (lioncash) [inline Source/Core/Common/HandShake.h]: Our coding style puts the public interface first.

File: Source/Core/Common/HandShake.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    HandShake& parent;
+    std::unique_lock<std::mutex> guard;
+
+  public:
+    // a concurrency guru should probably check whether these memory orders are correct
```
Comment (lioncash) [inline Source/Core/Common/HandShake.h]: You should probably know how to properly do this sort of thing if you're dealing with threading at all, just an FYI. This is, to me, a tremendous red flag for what's to come.

File: Source/Core/Common/HandShake.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                          std::memory_order_release);
+    }
+    Inner& GetRef() { return parent.sides[parent.select.load(std::memory_order_relaxed)].inner; }
+  };
+  // dropping a value of this type will release the other side to the reader
+  class YieldGuard
+  {
+    friend class HandShake<Inner>;
+
+  private:
```
Comment (lioncash) [inline Source/Core/Common/HandShake.h]: Ditto about a public interface being first.

File: Source/Core/Common/HandShake.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    side ^= 1;
+    auto res = std::make_optional<YieldGuard>(*this);
+    writerGuard.swap(res->guard);
+    return res;
+  }
+
+private:
+  // depends on microarchitecture; usual values are 32, 64 or 128; this is used to avoid false
+  // sharing, so use the highest value in use on the given architecture
+  static constexpr int CACHELINE_SIZE = 128;
```
Comment (lioncash) [inline Source/Core/Common/HandShake.h]: `std::hardware_destructive_interference_size` already provides this on MSVC.

File: Source/Core/Common/HandShake.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  /// if we could detect whether a mutex is being waited on (theoretically possible in most mutex
+  /// implementations) we wouldn't need this
+  std::atomic<int> select = 0;
+  struct
+  {
+    // separate into different cache lines to avoid false sharing
+    alignas(CACHELINE_SIZE) std::mutex mutex;
+    Inner inner;
+  } sides[2];
```
Comment (lioncash) [inline Source/Core/Common/HandShake.h]: std::array

File: Source/Core/Core/PowerPC/Jit64/Jit.h [hunk 6]
```diff
@@ -36,6 +36,11 @@ struct CodeOp;
 
 class Jit64 : public Jitx86Base
 {
+  // these three need access to the JitState
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/Jit64/Jit.h]: This is self-evident by the use of the friend declarations

File: Source/Core/Core/PowerPC/JitTiered/TieredGeneric.cpp [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+  // let the subclass decide how to get a block
+  // (or whether to create an interpreter block instead)
+  return LookupBlock(key, address);
+}
+
+u32 JitTieredGeneric::LookupBlock(u32 key, u32 address)
+{
+  // we have no JIT to get blocks from, so just create new interpreter block unconditionally
+  disp_cache[key].address = address;
+  disp_cache[key].offset = (u32)inst_cache.size();
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitTiered/TieredGeneric.cpp]: static_cast

File: Source/Core/Core/PowerPC/JitTiered/JitTiered.h [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  static constexpr int VICTIM_SETS_SHIFT = 3;
+  static constexpr size_t VICTIM_SETS = 1 << VICTIM_SETS_SHIFT;
+  static constexpr int VICTIM_WAYS_SHIFT = 5;
+  static constexpr size_t VICTIM_WAYS = 1 << VICTIM_WAYS_SHIFT;
+  static constexpr size_t DISP_CACHE_SIZE = DISP_PRIMARY_CACHE_SIZE + VICTIM_SETS * VICTIM_WAYS;
+
+  /// direct-associative cache (hash table) for blocks, followed by victim cache
+  /// (contiguous for iteration convenience)
+  /// JIT threads may only delete entries, not create them (safety invariant)
+  DispCacheEntry disp_cache[DISP_CACHE_SIZE]{};
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitTiered/JitTiered.h]: std::array

File: Source/Core/Core/PowerPC/JitTiered/JitTiered.h [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+  /// second-chance bits for the WS-Clock eviction algorithm
+  std::bitset<VICTIM_SETS * VICTIM_WAYS> victim_second_chance;
+  /// clocks for the WS-Clock eviction algorithm
+  u8 victim_clocks[VICTIM_SETS];
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitTiered/JitTiered.h]: std::array

File: Source/Core/Core/PowerPC/JitTiered/TieredGeneric.cpp [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+#include "Core/PowerPC/MMU.h"
+#include "Core/PowerPC/PowerPC.h"
+
+static constexpr u32 EXCEPTION_SYNC =
+    ~(EXCEPTION_EXTERNAL_INT | EXCEPTION_PERFORMANCE_MONITOR | EXCEPTION_DECREMENTER);
+
+void JitTieredGeneric::ClearCache()
+{
+  // invalidate dispatch cache
+  std::memset(&disp_cache, 0, sizeof(disp_cache));
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitTiered/TieredGeneric.cpp]:
```cpp
disp_cache = {};
```

File: Source/Core/Common/HandShake.h (hunk 2 above)
Comment (CrystalGamma) [reply Source/Core/Common/HandShake.h]: I *have* thought about all ways (that I can think of) how this interaction can play out for a few hours, I just haven't gotten to having two threads use this yet (real-world systems might not behave like my understanding of the memory model, after all). I hope not to need more than this primitive and a handful of mutexes anyway, so there is not really 'more to come' except considerations for I$ sync and deadlocking.

File: Source/Core/Common/HandShake.h (hunk 4 above)
Comment (CrystalGamma) [reply Source/Core/Common/HandShake.h]: cppreference.com tells me that is C++17. How should I handle this to have it portable?

File: Source/Core/Core/PowerPC/Jit64/Jit.h (hunk 6 above)
Comment (CrystalGamma) [reply Source/Core/Core/PowerPC/Jit64/Jit.h]: I'm pretty sure this is part of #7602 …
Anyway, I put this comment there to make clear that it is JitState (and not some other member) that is accessed by these classes.

File: Source/Core/Common/HandShake.h [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  Inner& GetWriter() { return sides[side].inner; }
+  std::optional<YieldGuard> Yield()
+  {
+    if (select.load(std::memory_order_acquire) != side)
+    {  // reader is still on the other side
+      return {};
+    }
+    side ^= 1;
+    auto res = std::make_optional<YieldGuard>(*this);
+    writerGuard.swap(res->guard);
```
Comment (MerryMage) [inline Source/Core/Common/HandShake.h]: I am finding the logic behind this hard to understand. 

You create a `lock_guard` in `YieldGuard`, then hand it off to `HandShake`, then never release it until `HandShake` is destructed.

This doesn't seem to have a purpose.

**Comment by CrystalGamma** [comment]
Let me try to explain the intention behind HandShake: one end (the "Writer", though actually, both sides have write access to their side) can always write and read their side and never block on the handshake (because my goal is to never have the CPU thread stop on anything except cleaning up the dispatch cache on code space reclamation, and even then only if it is too fast doing other things) while the other side (the "Reader") only accesses its side around 50% of the time and may block. Thus writerGuard is never released (but potentially swapped in Yield).

Meanwhile, there are two main reasons for not using a standard message queue: 1) my plans for invalidation require the CPU thread to know if the previous invalidations have been processed by the JIT thread (the CPU thread only manages its dispatch cache) and 2) non-blocking message queues may have backpressure and thus latency problems, aside from needing to allocate every message.

File: Source/Core/Common/HandShake.h (hunk 11 above)
Comment (CrystalGamma) [reply Source/Core/Common/HandShake.h]: I don't see how that would deadlock, since the method returns early if select hasn't changed.

Comment (MerryMage) [reply Source/Core/Common/HandShake.h]: That was my error, I apologise.

Sequence of events:

1. On the first call to `Yield`:
   1. `sides[0].mutex` is acquired by `lock_guard` in `YieldGuard`'s constructor on line 20.
   1. the guard is swapped with `writerGuard`
   1. `Yield` returns. We note that we have currently holding an active lock on `sides[0].mutex`.
2. Subsequent calls to `Yield` return `std::nullopt`.
3. A swap happens.
1. On the next call to `Yield`:
   1. `sides[1].mutex` is acquired by `lock_guard` in `YieldGuard`'s constructor.
   1. the guard is swapped with `writerGuard`
   1. `Yield` returns. We note that we have currently holding an active lock on `sides[0].mutex` and `sides[1].mutex`.
1. `sides[0].mutex` is eventually released when `YieldGuard` is destructed.
1. Repeat from 2 ad nauseum.

**Comment by CrystalGamma** [comment]
I suppose just duplicating the report data structures (using try_lock on the CPU thread and keeping the mutex locked as long as the JIT thread does not want data, and using a CV to wake it back up once the CPU thread has written its report) might do the job though, since (mostly in code not included in this PR) the handoff sequence copies things around anyway. Thanks for reminding me that my use case is not as special as I thought :S

**Comment by jordan-woyak** [comment]
It's been some years and the requested changes have not been made. It also needs a rebase. I'm closing this for now, but it can be re-opened upon request.

**Pull request closed without merging**